proxy/context.go

```diff
@@ -233,10 +237,104 @@ func (c *context) clone() *context {
 	return &cc
 }
 
+func (c *context) wasExecuted()bool {
+	return c.executionCounter != 0
+}
+
 func (c *context) setMetricsPrefix(prefix string) {
 	c.metrics.prefix = prefix + ".custom."
 }
 
+func (tt *teeTie) Read(b []byte) (int, error) {
+	n, err := tt.r.Read(b)
+
+	if err != nil && err != io.EOF {
+		tt.w.CloseWithError(err)
+		return n, err
+	}
+
+	if n > 0 {
+		if _, werr := tt.w.Write(b[:n]); werr != nil {
+			log.Error("tee: error while tee request", werr)
+		}
+	}
+
+	if err == io.EOF {
+		tt.w.Close()
+	}
+
+	return n, err
+}
+
+func (tt *teeTie) Close() error { return nil }
+
+// Returns the cloned request and the tee body to be used on the main request.
+func cloneRequest(u *url.URL, req *http.Request) (*http.Request, io.ReadCloser, error) {
+	h := make(http.Header)
+	for k, v := range req.Header {
+		h[k] = v
+	}
+
+	var teeBody io.ReadCloser
+	mainBody := req.Body
+
+	// see proxy.go:231
+	if req.ContentLength != 0 {
+		pr, pw := io.Pipe()
+		teeBody = pr
+		mainBody = &teeTie{mainBody, pw}
+	}
+
+	clone, err := http.NewRequest(req.Method, u.String(), teeBody)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clone.Header = h
+	clone.ContentLength = req.ContentLength
+
+	return clone, mainBody, nil
+}
+
+func (c *context) Split() (filters.FilterContext, error) {
+	originalRequest := c.Request()
```

Comment:
we should preserve the opentracing server span, if available. Sg like (pseudo):

```
serverSpan := opentracing.SpanFromContext(originalRequest.Context())
cr = cr.WithContext(opentracing.ContextWithSpan(cr.Context(), otspan))
```

this way the cloned request's stdlib context and the original request's stdlib is not tied together, but both of them have the same server span, and the opentracing graph will show the tee request nicely.

Reply:
Is SpanFromContext cloning the context or returning a reference? If it is a reference I assume the span to be closed in the main request. If not, should I close the cloned version of the span?

Reply:
I pushed this change. Could you review it?

Reply:
lgtm